lib/tcl-duktape.c

```diff
@@ -93,6 +110,209 @@ parse_id(ClientData cdata, Tcl_Interp *interp, Tcl_Obj *const idobj, int del)
     return ctx;
 }
 
+/*
+ * Deal with Duktape Lambdas using a custom Tcl Obj type which
+ * we can use to free the lambda when the object goes away
+ */
+/**
+ ** Free a Duktape lambda object when the Tcl one is freed
+ ** also free all the internal mechanisms required to get
+ ** this far
+ **/
+static void Tclduk_LambdaObjType_Free(Tcl_Obj *lambdaObj) {
+    struct DuktapeLambdaInstanceData *instanceData;
+    ClientData cdata;
+    Tcl_Obj *handle, *lambdaNameObj, *bytecode;
+    const char *lambdaName;
+    int lambdaNameLength;
+    duk_context *ctx;
+    int freeDukLambda;
+
+    instanceData = lambdaObj->internalRep.otherValuePtr;
+
+    freeDukLambda = 1;
+
+    instanceData->refCount--;
+
+    if (instanceData->refCount > 0) {
+        return;
+    }
+
+    cdata         = instanceData->cdata;
+    lambdaNameObj = instanceData->lambdaName;
+    handle        = instanceData->handle;
+    bytecode      = instanceData->bytecode;
+
+    ckfree(instanceData);
+
+    Tcl_DecrRefCount(bytecode);
+
+    /*
+     * Find the Duktape heap context from the handle
+     */
+    ctx = parse_id(cdata, NULL, handle, 0);
+    Tcl_DecrRefCount(handle);
+    if (ctx == NULL) {
+        Tcl_DecrRefCount(lambdaNameObj);
+        return;
+    }
+
+    /*
+     * Schedule the deletion of the lambda from Duktape
+     */
+    if (freeDukLambda) {
+        lambdaName = Tcl_GetStringFromObj(lambdaNameObj, &lambdaNameLength);
+        duk_push_global_stash(ctx);                                  /* => ... [stash] */
+        duk_get_prop_literal(ctx, -1, "freeableLambdas");            /* => ... [stash] [object|undefined] */
+        if (duk_is_undefined(ctx, -1)) {
+            duk_pop(ctx);                                            /* => ... [stash] */
+            duk_push_object(ctx);                                    /* => ... [stash] [object] */
+            duk_dup(ctx, -1);                                        /* => ... [stash] [object] [object] */
+            duk_put_prop_literal(ctx, -3, "freeableLambdas");        /* => ... [stash.freeableLambdas=object] [object] */
+        }
+        duk_push_true(ctx);                                          /* => ... [stash] [object] [true] */
+        duk_put_prop_lstring(ctx, -2, lambdaName, lambdaNameLength); /* => ... [stash] [object.lambdaName=true] */
+        duk_pop_2(ctx);                                              /* => ... */
+    }
+
+    /*
+     * Finish cleanup
+     */
+    Tcl_DecrRefCount(lambdaNameObj);
+
+    return;
+}
+
+static void Tclduk_LambdaObjType_Dup(Tcl_Obj *src, Tcl_Obj *dest) {
+    struct DuktapeLambdaInstanceData *instanceData;
+
+    instanceData = src->internalRep.otherValuePtr;
+
+    instanceData->refCount++;
+
+    dest->internalRep.otherValuePtr = instanceData;
+
+    return;
+}
+
+static void Tclduk_LambdaObjType_String(Tcl_Obj *lambdaObj) {
```

Comment:
String serialization needs a test.

Reply:
The test for string serialization is in the lambda test -- when a JavaScript/Duktape lambda is passed to `tcleval()` in the test, it has to be serialized to a string first.

Reply:
Sorry, I didn't go into enough detail.  I meant a test that compares the string representation of a lambda (like below) against a reference.  However, this test will end up either brittle (since the bytecode is allowed to change in minor releases) or so broad it isn't very useful (if you match against a lax regexp/glob pattern).  I'm dropping this.  If I come up with a good serialized closure test I'll add it myself later.

```tcl
{apply {args {set code vwAAAAQAAAABAAAAAAABAAAAAAAAAAAAABgEC4AAAAB+AAAACwAAAJ0AAACeAAAAAAVjb3VudAAAAAAAAAAAAAAABGV2YWwAAAANBAAAAAEAAAAMAAAAAAAAAAD/////
set lambdaName lambda_20448977631102520059783368690_0
set handle ::duktape::14
return [::duktape::evallambda $handle $code $lambdaName {*}$args]}}}
```